ships/0028-buildrun-cleanup.md

```diff
@@ -0,0 +1,152 @@
+- If the number of successful BuildRuns is larger than `successfulLimit`, then the oldest successful BuildRun(s) sorted by `.metadata.creationTimestamp` will be deleted.
+- If the number of failed BuildRuns is larger than `failedLimit`, then the oldest failed BuildRun(s) sorted by `.metadata.creationTimestamp` will be deleted.
+
+We can consider to run above logic in a go routine rather than as part of the BuildRun reconciliation. Failures would be ignored which is okay.
+
+When the `ttlAfterFailed` field is set for a BuildRun that failed, or when the `ttlAfterSuccessful` field is set for a BuildRun that succeeded, then we will return a reconciliation result where `Requeue` is set to `true` and `RequeueAfter` is set to `.status.completionTime + TTL - now`.
+
+When a controller becomes the leader, a `CreateEvent` happens for all BuildRuns. [Today, we only reconcile](https://github.com/shipwright-io/build/blob/v0.8.0/pkg/reconciler/buildrun/controller.go#L57-L63) if the BuildRun has not yet a TaskRun reference and has not yet completed. We will also need to trigger a reconcile in case it has completed and there is a TTL for the BuildRun.
+
+Our BuildRun reconciliation logic will need to be extended. [After we fetched the BuildRun](https://github.com/shipwright-io/build/blob/v0.8.0/pkg/reconciler/buildrun/buildrun.go#L82), we will need to add logic that checks if the BuildRun is already completed and check if we need to delete it or requeue for deletion.
+
+TBD When a Build gets modified and any value inside the `retention` section gets initially set or lowered, then ... [see planned discussion in community meeting](https://github.com/shipwright-io/community/issues/62#issuecomment-1041378397).
+
+### Staging
+
+The implementation can be done in stages. In particular, the TTL-based cleanup and the limit-based cleanup are very much independent.
+
+### Documentation updates
+
+We will need to update the documentation about builds to introduce this new feature.
+
+### User Stories
+
+#### As a Build user, I want to define how many finished BuildRuns I want to keep for my Build so that the amount of finished BuildRuns does not grow endlessly
+
+#### As a Build user, I want to define how long a finished BuildRun should stay so that I do not have obsolete and very old BuildRuns in the system
+
+### Test Plan
+
+It should include:
+
+- Integration tests
+- Unit tests
+- e2e tests
+
+### Release Criteria
+
+Should be available before release v1.0.0
+
+### Risks and Mitigations
+
+We will have more logic in our reconciliations. Especially the required `list` calls can become a problem if `successfulLimit`, or `failedLimit` are set to high values and we consequently retrieve a large amount of BuildRuns from the API server. A mitigation would be that we define maximum values for those values (maybe 100).
+
+## Drawbacks
+
+None.
+
+## Alternatives
```

Comment:
Can we consider having a separate controller? I need to spend more time thinking on the pros/cons of this.

Reply:
Yep, I had that idea as well. Need to think about this again now that we discussed how we handle Build updates.

Reply:
ok, let me try to brainstorm today/tomorrow, before putting my thoughts here. Thanks.

Reply:
Adding here my thoughts(already expressed them offline to Sascha):

My concern on not using another controller, is in terms of the reconcile loop increase. My rational in terms of reconciles is the following(based on the worst scenario on ttl):

Assuming 600 completed Buildruns with TTL. 
- Today a single BR reconciles 4-5 times only. => 3000 reconciles
- We increase per BR +1 reconcile(due to the requeue)  => +600 reconciles
- On rollouts of the controller,  assuming all 600 are completed and waiting for a requeue(worst scenario), thats a +600 (we didnt do this before) => +600 reconciles
- In total +1200, a 40% plus. Significant increase.

An alternative for the rollouts would be introducing rate limiting in the controller, see [ratelimiter](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/ratelimiter#RateLimiter) that might help with one of the extra 600.

Reply:
Some quick "prior art" for reference.  OpenShift build pruning has handled things in "the same" controller / reconciler loop for many years with little to no problems, but OpenShift builds does not have TTL based pruning, only count/total based.

It will be interesting to see as we work the implementation which of the following approaches turn out most palatable:
- count based pruning in the existing controller/reconciler loop, but ttl based in a new controller/reconciler
- both count and ttl based in a new controller/reconciler
- both count and ttl based in the existing controller/reconciler

Seeing this also resulted in the following notion popping into my head:  do we need both TTL and count based pruning out of the gate?  i.e. is there some CodeEngine specific req wanting TTL?  Or is it sufficient to go out with say just count based pruning initially, then wait for feedback on whether that is sufficient?

FWIW, there have been no requests for TTL based pruning that I can recall from OpenShift Build customers (where OCP Builds was first released in 2015).

Reply:
Thanks @gabemontero for the feedback. The main motivation to have the TTL is related to the standalone BuildRuns where there is no Build (see the other ship from Baran). If there is no Build, there is no count-based pruning.

Reply:
ah thanks for the clarification @SaschaSchwarze0 ... yeah, I have not had a chance to look at that EP yet, and the title of `One-Off Builds` did not immediately connect with me that it meant "BuildRuns without Builds", but yes it makes sense, that without that organizational aspect of "being under a Build", the count based pruning does not make sense.  Or more precisely, the only option is "global" limits and that is not what is desired.

I'm wrapping up my review of the totality of the latest changes now.  Thanks.